File: algorithms/solutions/week33/manhdv/middle_of_the_linked_list.rb
# https://leetcode.com/problems/middle-of-the-linked-list/

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}
def middle_node(head)
  one_step_move_start = head
  two_step_move_start = head

  loop do
    break if two_step_move_start&.next.nil?
    two_step_move_start = two_step_move_start.next&.next
    one_step_move_start = one_step_move_start.next
  end

  one_step_move_start
end

File: algorithms/solutions/week33/taurandat/middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
  ListNode* middleNode(ListNode* head) {
    ListNode* fast = head;
    ListNode* slow = head;
    while(fast != NULL && fast->next != NULL) {
      slow = slow->next;
      fast = fast->next->next;
    }
    return slow;
  }
};
